tests/run/i9011.scala

```diff
@@ -0,0 +1,57 @@
+enum Opt[+T] derives Eq:
+  case Sm(t: T)
+  case Nn
+
+import scala.deriving._
+import scala.compiletime.{erasedValue, summonInline}
+
+trait Eq[T] {
+  def eqv(x: T, y: T): Boolean
+}
+
+object Eq {
+  given Eq[Int] {
+    def eqv(x: Int, y: Int) = x == y
+  }
+
+  inline def summonAll[T <: Tuple]: List[Eq[_]] = inline erasedValue[T] match {
+    case _: Unit => Nil
+    case _: (t *: ts) => summonInline[Eq[t]] :: summonAll[ts]
+  }
+
+  def check(elem: Eq[_])(x: Any, y: Any): Boolean =
+    elem.asInstanceOf[Eq[Any]].eqv(x, y)
+
+  def iterator[T](p: T) = p.asInstanceOf[Product].productIterator
```

Comment:
Going back to the original issue, is this cast actually safe?

> The example in the docs suggests that if we have a Mirror.ProductOf[T] then we can safely cast an instance of T to Product (in order to access productIterator etc.). But this is not the case.

If this isn't actually guaranteed, then the documentation needs to be updated to not rely on this cast in an example.